exercises/react/react.ts

```diff
@@ -0,0 +1,102 @@
+//
+// Delete and replace stub with your own implementation
+//
+// Inspired by "How it Works":
+// https://indepth.dev/posts/1269/finding-fine-grained-reactive-programming#how-it-works
+// https://levelup.gitconnected.com/finding-fine-grained-reactive-programming-89741994ddee?source=friends_link&sk=31c66a70c1dce7dd5f3f4229423ad127#4543
+//
+// and "Computations":
+// https://github.com/ryansolid/solid/blob/master/documentation/reactivity.md#user-content-computations
+//
+
+type EqualFn<T> = (lhs: T, rhs: T) => boolean
+type GetterFn<T> = () => T
+type SetterFn<T> = (v: T) => T
+type UnsubscribeFn = () => void
+type UpdateFn<T> = (value?: T) => T
+
+type InputCouple<T> = [GetterFn<T>, SetterFn<T>]
+
+type Options = {
+  name: string // for debugging
+}
+
+type ObserverR = {
+  name?: string
+}
+
+type ObserverV<T> = {
+  value?: T
+  fn: UpdateFn<T>
+}
+
+type Observer<T> = ObserverR & ObserverV<T>
+
+type SubjectR = {
+  name?: string
+  observer: ObserverR | undefined
+}
+
+type SubjectV<T> = {
+  value: T
+  equal?: EqualFn<T>
+}
+
+type Subject<T> = SubjectR & SubjectV<T>
+
+// module Context value
+let activeObserver: ObserverR
+
+function updateObserver<T>(observer: Observer<T>): void {
+  const prevObserver = activeObserver
+  activeObserver = observer
+  observer.value = observer.fn(observer.value)
+  activeObserver = prevObserver
+}
+
+function createInput<T>(
+  value: T,
+  _equal?: boolean | EqualFn<T>,
+  options?: { name?: string }
+): InputCouple<T> {
+  const s: Subject<T> = {
+    name: options?.name,
+    observer: undefined,
+    value,
+    equal: undefined,
+  }
+
+  const read: GetterFn<T> = () => {
+    if (activeObserver) s.observer = activeObserver
+    return s.value
+  }
+
+  const write: SetterFn<T> = (value) => {
+    s.value = value
+    if (s.observer) updateObserver(s.observer as Observer<unknown>)
+    return s.value
+  }
+
+  return [read, write]
+}
+
+function createComputed<T>(
+  fn: UpdateFn<T>,
+  value?: T,
+  _equal?: boolean | EqualFn<T>,
+  options?: { name?: string }
+): GetterFn<T> {
+  const o: Observer<T> = {
+    name: options?.name,
+    value,
+    fn,
+  }
+  updateObserver(o)
+  return (): T => o.value!
+}
+
```

Comment:
Is this meant to be in the _stub_? If yes (and I think yes), let's point out _where_ someone needs to implement something!

Reply:
At the beginning of the file:
```typescript
// Delete and replace stub with your own implementation
```
So it's not a "fill in the blanks" type of situation. 

* This stub contains the just enough functionality to pass the `compute cells update value when inputs are changed` test case. It you compare `createComputed` to the reference implementation you should notice that the stub creates an `Observer<T>` while the reference creates a `Computed<T>`.

* It clearly identifies the API and types that need to be implemented.

* The code contains bits and pieces to defuse some showstopper frustrations for less determined students (we're all tired at the end of the day). In particular the `ObserverV<T>`, `ObserverR` split - plus `unknown` for `write` to dodge the explicit `any` bullet.

It's supposed to credibly demonstrate that this isn't an impossible challenge (in comparison to the JavaScript one where dependencies are explicitly passed in addition to the function) and showing `activeObserver` is supposed to make the _magic_ go away.

Reply:
> It's supposed to credibly demonstrate that this isn't an impossible challenge (in comparison to the JavaScript one where dependencies are explicitly passed in addition to the function) and showing activeObserver is supposed to make the magic go away.

I like this. I think we shoud have that as a comment below "// Delete and replace stub with your own implementation"